Don't test forward compatibility in qpy compat tests

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

The qpy compatibility tests are designed to test backwards compatibility. They generate QPY files using all historical releases of Qiskit with QPY support and test that all proposed changes are able to load the qpy files. There was an underlying assumption to how the tests were written that the version under test is always newer than all the historical releases. This has been true in the past because the release model for qiskit mean that we were maintaining at most one stable release at a time, and when we released a new minor version of qiskit we retired the previous minor version release series's branch. However, moving forward this assumption will no longer be true as we're moving to a model where we support > 1 branch at a time. When we start supporting > 1 release series at a time, PRs proposed to the older stable branches will no longer be able to test all releases because QPY only support backwards compatibility and will not be able to load QPY files generated from newer releases. The example of this was in the 0.25.x series, there is overlap in the support of the stable/0.25 branch and the stable/0.45 branch, and after the release of 0.45.0rc1 the tests started to fail because you can't load QPY files generated with 0.45.0rc1 with 0.25.x.

This commit address this by adding a condition to the QPY compat scripts to ensure that the version under test is >= the QPY file source version. If we encounter that scenario we just skip the tests. This should prevent this situation from occurring for future releases as we expand our support windows.

### Details and comments

mtreinish committed:
Don't test forward compatibility in qpy compat tests

The qpy compatibility tests are designed to test backwards
compatibility. They generate QPY files using all historical releases
of Qiskit with QPY support and test that all proposed changes are
able to load the qpy files. There was an underlying assumption to how
the tests were written that the version under test is always newer
than all the historical releases. This has been true in the past because
the release model for qiskit mean that we were maintaining at most one
stable release at a time, and when we released a new minor version of
qiskit we retired the previous minor version release series's branch.
However, moving forward this assumption will no longer be true as we're
moving to a model where we support > 1 branch at a time. When we start
supporting > 1 release series at a time, PRs proposed to the older
stable branches will no longer be able to test all releases because QPY
only support backwards compatibility and will not be able to load QPY
files generated from newer releases. The example of this was in the
0.25.x series, there is overlap in the support of the stable/0.25 branch
and the stable/0.45 branch, and after the release of 0.45.0rc1 the tests
started to fail because you can't load QPY files generated with
0.45.0rc1 with 0.25.x.

This commit address this by adding a condition to the QPY compat scripts
to ensure that the version under test is >= the QPY file source version.
If we encounter that scenario we just skip the tests. This should
prevent this situation from occurring for future releases as we expand
our support windows.

mtreinish commented on test/qpy_compat/process_version.sh:
```diff
@@ -14,9 +14,23 @@
+# qiskit_version is the version under test, We're testing that we can correctly
+# read the qpy files generated with source version with this version.
+qiskit_version=`../qiskit_venv/bin/python -c "import qiskit;print(qiskit.__version__`)"`
+qiskit_version_pars = ${qiskit_version//./ } )
```
```suggestion
qiskit_version_pars = ( ${qiskit_version//./ } )
```

mtreinish commented on test/qpy_compat/process_version.sh:
```diff
@@ -14,9 +14,23 @@
-if [[ ${parts[1]} -lt 18 ]] ; then
+# qiskit_version is the version under test, We're testing that we can correctly
+# read the qpy files generated with source version with this version.
+qiskit_version=`../qiskit_venv/bin/python -c "import qiskit;print(qiskit.__version__`)"`
```
```suggestion
qiskit_version=`../qiskit_venv/bin/python -c "import qiskit;print(qiskit.__version__)"`
```

mtreinish committed:
Fix syntax errors

jakelishman commented on test/qpy_compat/process_version.sh:
```diff
@@ -14,9 +14,23 @@
+# qiskit_version is the version under test, We're testing that we can correctly
+# read the qpy files generated with source version with this version.
+qiskit_version=`../qiskit_venv/bin/python -c "import qiskit;print(qiskit.__version__)"`
+qiskit_version_pars = ( ${qiskit_version//./ } )
```
There's a naming problem here - we call the variable `qiskit_version_pars` (lol) but then use it as `qiskit_parts`.

mtreinish replied:
I fixed this in: https://github.com/Qiskit/qiskit/pull/11101/commits/1c87c502c228942723fd7299af648881687de35e (I should have split it out into a second commit but forgot and just did `git add` without the `-p`)

jakelishman commented on test/qpy_compat/process_version.sh:
```diff
@@ -14,9 +14,23 @@
+# If the source version is newer than the version under test exit fast because
+# there is no QPY compatibility for loading a qpy file generated from a newer
+# release with an older release of Qiskit.
+if [[ ${parts[0]} -gt ${qiskit_parts[0]} || ${parts[1]} -gt ${qiskit_parts[1]} ]] ; then
```
This doesn't implement tuple comparison quite right - the right-hand side of the `||` needs to limit to the case that the minor versions are _equal_ (i.e. forbid `${parts[0]} -lt ${qiskit_parts[1]}`).

If we don't want to get into the weeds on doing that properly in bash, we could do something like

```bash
if [[ $(printf '%s\n%s\n' "$version" "$qiskit_version" | sort -V | head -n1) == $version]]
```

(or whatever it is) and let `sort -V` do the ordering for us.

mtreinish replied:
So I played with this locally and the only issue is if it's a pre-release `sort -V` doesn't handle this correctly it treats the suffix as a higher version. So if we're doing an upgrade test from 0.45.0rc1 to 0.45.0 it will evaluate that as trying to use a newer version here. I think I'm just going to write a small python script to do the comparison exactly as we want, rather than try to figure out how to best do this in bash.

jakelishman replied:
Oh yeah, now I remember hitting that in #10610, but it didn't matter there because prereleases would never be deployed, so I'd already filtered them out.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11101/commits/1c87c502c228942723fd7299af648881687de35e

mtreinish committed:
Merge remote-tracking branch 'origin/main' into fix-qpy-backwards-compat

mtreinish committed:
Use a Python script to compare versions

This commit adds a small python script to compare two version strings
and determine whether the version under test is newer (or the same) as
the source version. The logic for making this determination was a bit
more involved in bash, so using a python script lets us reuse the regex
built into qpy's interface to support this comparison already.

jakelishman commented on test/qpy_compat/process_version.sh:
```diff
@@ -20,17 +20,20 @@ version=$1
-if [[ ${parts[0]} -gt ${qiskit_parts[0]} || ${parts[1]} -gt ${qiskit_parts[1]} ]] ; then
+if [[ $valid_version -ne 0 ]] ; then
     exit 0
 fi
```
We `set -e`, so line 26 will error out with a failure when it should be a successful "do nothing".  Do we want something more like
```bash
if ! python compare_versions.py "$version" "$qiskit_version"; then
   exit 0
fi
```
?

mtreinish replied:
Oh, yeah that's a good call, all I tested was that it didn't build the venv and didn't think to look at the return code from the script. I'll make this change.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/11101/commits/2c8526d1ab472238dc0f75a1d7502702492b1f63

mtreinish committed:
Fix lint

mtreinish committed:
Fix return code with invalid version

This commit fixes the return code generation if an invalid comparison is
requested. We want to handle this with a 0 return code to treat it as a
skip, but the previous logic would error before we evaluated the if
statement.

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

jakelishman reviewed (approved):
This seems right to me now.  I suppose it's technically possible that if we ever cause the bash script to feed the wrong inputs to the Python script we'll start skipping every version of QPY (unhandled exception has nonzero return code), but that'll at least show something in the logs.
